File: tests/generation/test_bsp.py
"""Tests all functions in generation/bsp.py."""
from __future__ import annotations

# Builtin
from typing import TYPE_CHECKING

# Custom
from game.generation.bsp import Leaf
from game.generation.primitives import Point

if TYPE_CHECKING:
    import numpy as np

__all__ = ()


def test_bsp_init(
    boundary_point: Point, invalid_point: Point, grid: np.ndarray
) -> None:
    """Test the initialisation of the Leaf class in bsp.py.

    Parameters
    ----------
    boundary_point: Point
        A boundary point used for testing.
    invalid_point: Point
        An invalid point used for testing.
    grid: np.ndarray
        The 2D grid used for testing.
    """
    assert (
        repr(Leaf(boundary_point, Point(grid.shape[1], grid.shape[0]), grid))
        == "<Leaf (Left=None) (Right=None) (Top-left position=Point(x=0, y=0))"
        " (Bottom-right position=Point(x=50, y=50))>"
    )
    assert (
        repr(Leaf(invalid_point, Point(grid.shape[1], grid.shape[0]), grid))
        == "<Leaf (Left=None) (Right=None) (Top-left position=Point(x='test',"
        " y='test')) (Bottom-right position=Point(x=50, y=50))>"
    )


def test_leaf_split(leaf: Leaf) -> None:
    """Test the split function in the Leaf class.

    Parameters
    ----------
    leaf: Leaf
        The leaf used for testing.
    """
    # Repeat until we have tested the vertical and horizontal split routes
    test_vertical = test_horizontal = False
    while not test_vertical or not test_horizontal:
        leaf.split()
        if leaf.split_vertical:
            test_vertical = True
        else:
            test_horizontal = True
        assert (
            leaf.left is not None
            and leaf.right is not None
            and isinstance(leaf.split_vertical, bool)
        )
        leaf.left = leaf.right = leaf.split_vertical = None


def test_leaf_create_room(leaf: Leaf) -> None:
    """Test the create_room function in the Leaf class.

    Parameters
    ----------
    leaf: Leaf
        The leaf used for testing.
    """
    # Repeat until a room is created since the ratio may be wrong sometimes
    result = leaf.create_room()
    while not result:
        result = leaf.create_room()
    assert result and leaf.room

    # Make sure we test what happens if the leaf's left and right nodes are not None
    leaf.left = leaf.right = "test"
    assert not leaf.create_room()

File: cpp_src/test/test_bsp.cpp
// External includes
#include "gtest/gtest.h"

// Custom includes
#include "bsp.hpp"
#include "fixtures.hpp"

// ----- TESTS ------------------------------
TEST_F(Fixtures, TestBspSplitCorrect) {
  // Split the bsp normally
  leaf.split(grid, random_generator, false);

  // Calculate the difference between the bottom right point of the left leaf
  // and the top right point of the right leaf, so we can determine it's split
  // direction
  Point leaf_diff = leaf.left->container.bottom_right - leaf.right->container.top_left,
      target_diff = (leaf_diff.x == 2) ? Point{2, 19} : Point{19, 2};

  // Test if the child leafs border each other
  ASSERT_EQ(leaf_diff, target_diff);
}

TEST_F(Fixtures, TestBspSplitDebug) {
  // Split the grid in debug mode
  leaf.split(grid, random_generator, true);

  // Find the row/column where the split occurred and get the grid position of the split
  int debug_pos_index = (int) (std::find(grid.grid.begin(), grid.grid.end(), TileType::DebugWall) - grid.grid.begin());
  Point debug_pos = {debug_pos_index % grid.width, debug_pos_index / grid.width};

  // Make sure the split runs the entire length of the grid. We need to
  // determine if the split was vertical or horizontal first however
  if (debug_pos.x > 0) {
    for (int y = 0; y < grid.width; y++) {
      ASSERT_EQ(grid.get_value({debug_pos.x, y}), TileType::DebugWall);
    }
  } else if (debug_pos.y > 0) {
    for (int x = 0; x < grid.height; x++) {
      ASSERT_EQ(grid.get_value({x, debug_pos.y}), TileType::DebugWall);
    }
  }
}

TEST_F(Fixtures, TestBspSplitSmallWidthHeight) {
  // Make sure we test what happens if the container's width and height are both
  // less than MIN_CONTAINER_SIZE
  leaf.container = Rect{Point{-1, -1}, Point{-1, -1}};
  ASSERT_FALSE(leaf.split(grid, random_generator, false));
}

TEST_F(Fixtures, TestBspCreateRoomChildLeaf) {
  // Repeat until a room is created since the ratio may be wrong sometimes then
  // test that the room is not null
  while (!leaf.create_room(grid, random_generator)) {}
  ASSERT_TRUE(leaf.room != nullptr);
}

TEST_F(Fixtures, TestBspCreateRoomNotNullLeftRight) {
  // Test what happens if the leaf and right leafs are not null
  Leaf temp_leaf = Leaf{Rect{Point{0, 0}, Point{0, 0}}};
  leaf.left = leaf.right = &temp_leaf;
  ASSERT_FALSE(leaf.create_room(grid, random_generator));
}
